### DISCUSSION TITLE
How to pass a custom config to overwrite the configs to train a model

### DISCUSSION THREAD

**Discussion by shadeekam** [open-edge-platform/anomalib#1888] [Q&A] [answered]
### Describe the bug

I am trying to train a model using  `anomalib train` where i have my own configurations for training and model cli is giving errors saying my configurations are wrong even though i used the configs from `--print-config`

### Dataset

N/A

### Model

PADiM

### Steps to reproduce the behavior

1. install anomalib
2. create custom configs for training , model , dataset
3. run following command `anomalib train -c ../configs/train_all.yaml --model ../configs/model.yaml`

### OS information

OS information:
- OS: [macOs 14.1]
- Python version: [3.11]
- Anomalib version: [source checked out ]
- I'm using a custom dataset
- Mac M3 apple silicon processor

### Expected behavior

should be able to train the model

### Screenshots

<img width="1489" alt="Screenshot 2024-02-13 at 22 11 33" src="https://github.com/openvinotoolkit/anomalib/assets/106906348/a3f51a82-bbce-4331-a91e-f1354ef7f3f1">

### Pip/GitHub

GitHub

### What version/branch did you use?

main

### Configuration YAML

```yaml
model:
  class_path: anomalib.models.Padim
  init_args:
    layers:
      - layer1
      - layer2
      - layer3
    input_size: null
    backbone: resnet18
    pre_trained: true
    n_features: null
```

### Logs

```shell
error: Parser key "model":
  Not a valid subclass of AnomalyModule. Got value: {'model': {'name': 'padim', 'init_args': {'layers': ['layer1', 'layer2', 'layer3'], 'input_size': None, 'backbone': 'resnet18', 'pre_trained': True, 'n_features': None}}}
  Subclass types expect one of:
  - a class path (str)
  - a dict with class_path entry
  - a dict without class_path but with init_args entry (class path given previously)
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by blueclowd** [comment]
Same here. Wondering if there is any runnable sample configuration on anomalib v1.x

**Comment by shadeekam** [comment]
Hi @blueclowd i was not able to run through terminal. But I trained a model from notebook, where i happen to key in all the params.

from anomalib.engine import Engine
from anomalib.utils.normalization import NormalizationMethod
```

engine = Engine(
    normalization=NormalizationMethod.MIN_MAX,
    threshold="F1AdaptiveThreshold",
    task=TaskType.SEGMENTATION,
    image_metrics=["AUROC"],
    accelerator="auto",
    check_val_every_n_epoch=1,
    limit_train_batches=1.0,
    limit_val_batches=1.0,
    limit_test_batches=1.0,
    precision=32,
    log_every_n_steps=50,
    devices=1,
    max_epochs=1,
    num_sanity_val_steps=0,
    val_check_interval=1.0,
)

engine.fit(model=model, datamodule=datamodule)
```

As I experienced certain configs formats are not yet updated in the updated source code

**Comment by blueclowd** [comment]
> Hi @blueclowd i was not able to run through terminal. But I trained a model from notebook, where i happen to key in all the params.
> 
> from anomalib.engine import Engine from anomalib.utils.normalization import NormalizationMethod
> 
> ```
> 
> engine = Engine(
>     normalization=NormalizationMethod.MIN_MAX,
>     threshold="F1AdaptiveThreshold",
>     task=TaskType.SEGMENTATION,
>     image_metrics=["AUROC"],
>     accelerator="auto",
>     check_val_every_n_epoch=1,
>     limit_train_batches=1.0,
>     limit_val_batches=1.0,
>     limit_test_batches=1.0,
>     precision=32,
>     log_every_n_steps=50,
>     devices=1,
>     max_epochs=1,
>     num_sanity_val_steps=0,
>     val_check_interval=1.0,
> )
> 
> engine.fit(model=model, datamodule=datamodule)
> ```
> 
> As I experienced certain configs formats are not yet updated in the updated source code

Thanks for sharing the hard work. Will try it out as well.

**Comment by Vbajeneza** [comment]
> ### Describe the bug
> I am trying to train a model using `anomalib train` where i have my own configurations for training and model cli is giving errors saying my configurations are wrong even though i used the configs from `--print-config`
> 
> ### Dataset
> N/A
> 
> ### Model
> PADiM
> 
> ### Steps to reproduce the behavior
> 1. install anomalib
> 2. create custom configs for training , model , dataset
> 3. run following command `anomalib train -c ../configs/train_all.yaml --model ../configs/model.yaml`
> 
> ### OS information
> OS information:
> 
> * OS: [macOs 14.1]
> * Python version: [3.11]
> * Anomalib version: [source checked out ]
> * I'm using a custom dataset
> * Mac M3 apple silicon processor
> 
> ### Expected behavior
> should be able to train the model
> 
> ### Screenshots
> <img alt="Screenshot 2024-02-13 at 22 11 33" width="1489" src="https://private-user-images.githubusercontent.com/106906348/304485719-a3f51a82-bbce-4331-a91e-f1354ef7f3f1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzOTQ2ODEsIm5iZiI6MTcxMDM5NDM4MSwicGF0aCI6Ii8xMDY5MDYzNDgvMzA0NDg1NzE5LWEzZjUxYTgyLWJiY2UtNDMzMS1hOTFlLWYxMzU0ZWY3ZjNmMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMxNFQwNTMzMDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02OTlmMDU0NmVlZTdjMjUxZWJiNDQwZDQxM2Y1YWE2YTk1OWU3MTViODdlMzIyNjZhZWMyMDYwM2FiYjJmZDUxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.1MTebVebO31t1C16sBeBZtDAR3sHatuh9aA1eflsL2A">
> ### Pip/GitHub
> GitHub
> 
> ### What version/branch did you use?
> main
> 
> ### Configuration YAML
> ```yaml
> model:
>   class_path: anomalib.models.Padim
>   init_args:
>     layers:
>       - layer1
>       - layer2
>       - layer3
>     input_size: null
>     backbone: resnet18
>     pre_trained: true
>     n_features: null
> ```
> 
> ### Logs
> ```shell
> error: Parser key "model":
>   Not a valid subclass of AnomalyModule. Got value: {'model': {'name': 'padim', 'init_args': {'layers': ['layer1', 'layer2', 'layer3'], 'input_size': None, 'backbone': 'resnet18', 'pre_trained': True, 'n_features': None}}}
>   Subclass types expect one of:
>   - a class path (str)
>   - a dict with class_path entry
>   - a dict without class_path but with init_args entry (class path given previously)
> ```
> 
> ### Code of Conduct
> * [x]  I agree to follow this project's Code of Conduct

Same issue here. Did you manage to resolve it?

**Comment by samet-akcay** [comment, marked as answer]
To pass a custom model or any component, you don't need to specify `model` section in the config file. These sections are only needed when a full config file is used. `--model` flag in the CLI already specifies that the yaml file will be for the model. So, to fix this you need to do the following:

```yaml
# ~~model~~: This is not needed when passed with --model flag from the CLI
class_path: anomalib.models.Padim
init_args:
  layers:
    - layer1
    - layer2
  input_size: null
  backbone: resnet18
  pre_trained: true
  n_features: null
```
